Enabling Server-side Tokenization, Prompt Generation, and Streamlined Benchmarking

stisiTT opened the pull request:
# Add Server-Side Tokenization Support for Cleaned-Random Dataset

## Summary

This PR adds a new `cleaned-random` dataset option to the vLLM benchmark serving script that leverages server-side tokenization for more accurate and stable prompt generation. The implementation follows a light-touch approach, preserving existing functionality while adding new capabilities.

## What's Changed

### New Files/Modifications

- **`vllm/benchmarks/benchmark_utils.py`**: Enhanced with server-side tokenization utilities
  - `SimplePromptClient`: Lightweight client for server communication
  - `generate_stable_prompt_tokens()`: Core tokenization logic
  - `generate_cleaned_random_prompts()`: High-level prompt generation
  - Tokenization helper functions for both client and server-side processing

- **`vllm/benchmarks/benchmarks_serving_old.py`**: Added `cleaned-random` dataset support
  - New dataset option in argument parser
  - Integration with server-side tokenization utilities
  - Health checking and error handling
  - Minimal changes to preserve existing functionality

## Key Features

### 🎯 **Server-Side Tokenization**
- Uses the actual vLLM server's tokenizer for precise token counting
- Eliminates client/server tokenization discrepancies
- Provides more accurate benchmark results

### 🧹 **Cleaned Random Prompts**
- Generates stable token sequences through encode→decode→encode cycles
- Removes tokenization artifacts and ensures consistency
- Reproducible results with seed control

### 🔧 **Light-Touch Integration**
- Preserves all existing functionality
- Follows established patterns in the codebase
- Minimal changes to core benchmark logic

## Usage

### Prerequisites
```bash
# Start vLLM server
vllm serve meta-llama/Llama-3.1-8B-Instruct \
    --host 127.0.0.1 \
    --port 8000 \
    --disable-log-requests

# Set authentication token
export OPENAI_API_KEY="test-key"
```

### Running Cleaned-Random Benchmarks
```bash
# Basic usage
python benchmarks/benchmarks_serving_old.py \
    --backend vllm \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name cleaned-random \
    --num-prompts 100 \
    --random-input-len 1024 \
    --random-output-len 128

# With custom request rate
python benchmarks/benchmarks_serving_old.py \
    --backend vllm \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --dataset-name cleaned-random \
    --num-prompts 500 \
    --random-input-len 512 \
    --random-output-len 256 \
    --request-rate 10 \
    --seed 42
```

## Technical Details

### Server-Side Tokenization Flow
1. **Health Check**: Verify server is accessible and responsive
2. **Token Generation**: Create random token sequences using server's vocab
3. **Cleaning Cycle**: Multiple encode→decode→encode passes for stability
4. **Prompt Creation**: Convert final tokens to text prompts
5. **Benchmark Execution**: Use cleaned prompts in standard benchmark flow

### Authentication
- Requires `OPENAI_API_KEY` environment variable
- Compatible with existing vLLM authentication patterns
- Graceful error handling for missing credentials

### Error Handling
- Server health validation before prompt generation
- Comprehensive error messages for debugging
- Fallback mechanisms where appropriate

## Benefits

### 🎯 **Accuracy**
- Eliminates tokenization mismatches between client and server
- More precise token counting for better benchmark metrics
- Consistent results across different environments

### 🔄 **Stability**
- Cleaned prompts remove tokenization artifacts
- Reproducible results with seed control
- Stable baseline for performance comparisons

### 🛠 **Maintainability**
- Light-touch implementation preserves existing code
- Modular design in `benchmark_utils.py`
- Easy to extend or modify

## Testing Considerations

### Manual Testing
```bash
# Test import functionality
python -c "from benchmarks.benchmark_utils import SimplePromptClient; print('✅ Import successful')"

# Test with running server
export OPENAI_API_KEY="test-key"
python benchmarks/benchmarks_serving_old.py \
    --dataset-name cleaned-random \
    --num-prompts 10 \
    --random-input-len 100 \
    --random-output-len 50 \
    --model meta-llama/Llama-3.1-8B-Instruct
```

### Verification Points
- [ ] Server health check works correctly
- [ ] Token generation produces stable sequences
- [ ] Prompt lengths match expected values
- [ ] Authentication errors are handled gracefully
- [ ] Existing benchmark functionality remains unchanged

## Backward Compatibility

✅ **Fully backward compatible**
- All existing dataset options work unchanged
- No modifications to core benchmark logic
- Optional feature that doesn't affect existing workflows

## Future Enhancements

- Add trace capture integration for performance optimization
- Support for multi-modal content in cleaned prompts
- Additional prompt cleaning strategies
- Performance metrics for tokenization overhead

## Related Issues

This implementation addresses the need for more accurate tokenization in benchmark scenarios while maintaining the simplicity and reliability of the existing codebase.

stisiTT committed:
feat(benchmarks): Add server-side tokenization and cleaned prompt generation

- Add new "cleaned-random" dataset option with server-side tokenization
- Introduce benchmark_utils package with CleanedPromptGenerator and PromptClient
- Enable accurate token counting using server's tokenizer instead of client-side approximation
- Add stable prompt generation through encode/decode cycles for reproducible benchmarks
- Maintain full backward compatibility with existing dataset options
- Follow vLLM's existing authentication patterns (OPENAI_API_KEY/AUTHORIZATION env vars)
- Graceful fallbacks for optional dependencies (torch -> numpy, missing utilities)

stisiTT committed:
Module Name Update

stisiTT committed:
Updating backend to handle old auth approach

stisiTT committed:
Commenting out redundant Single Run

tstescoTT commented on benchmarks/prompt_utils/prompt_client.py:
```diff
@@ -0,0 +1,180 @@
+logger.setLevel(logging.INFO)
+
+
+class PromptClient:
```
Please rename this class to not clash with the tt-inference-server PromptClient.

tstescoTT replied:
resolved

tstescoTT commented on benchmarks/prompt_utils/prompt_client.py:
```diff
@@ -0,0 +1,180 @@
+            logger.error(f"Server-side detokenization failed: {error_msg}")
+            return {"error": error_msg}
+
+    def capture_traces(self, context_lens: List[Tuple[int, int]], timeout: float = 1200.0) -> None:
```
Can you remove `capture_traces()`? We don't need it implemented here.

tstescoTT replied:
resolved

tstescoTT commented on benchmarks/benchmark_serving.py:
```diff
@@ -1083,6 +1229,21 @@ def main(args: argparse.Namespace):
+        "Requires CleanedPromptGenerator and PromptClient to be available."
+    )
+    
+    parser.add_argument(
```
`--disable-trace-capture` can be removed along with the trace capturing.

tstescoTT replied:
resolved

tstescoTT commented on benchmarks/benchmark_serving.py:
```diff
@@ -279,26 +383,30 @@ async def benchmark(
-            f"are correctly specified. Error: {test_output.error}")
-    else:
-        print("Initial test run completed. Starting main benchmark run...")
+    # COMMENTED OUT: Initial single prompt test run - redundant with capture_traces
```
We should keep the initial single prompt, we assume the traces have been captured but we want to have minimal edits to the benchmarking script.

tstescoTT replied:
resolved

tstescoTT commented on benchmarks/benchmark_serving.py:
```diff
@@ -57,9 +64,98 @@
 MILLISECONDS_TO_SECONDS_CONVERSION = 1000
 
 
+def setup_server_client(args, model_config):
```
Can setup_server_client() and generate_cleaned_random_requests() be added to the utils file?

tstescoTT replied:
resolved

tstescoTT commented on benchmarks/prompt_utils/cleaned_prompt_generator.py:
```diff
@@ -0,0 +1,227 @@
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Cleaned Prompt Generation Utility
```
"Cleaned" prompts make it seem like there was something wrong with the data which is confusing if we're saying it was random data. Can we call this "TokenizationTemplateLengthCorrected" or something that reflects whats happening?

tstescoTT replied:
resolved

stisiTT committed:
Refactoring for minimalist approach to server-side changes

Updated Commit

feat: Add server-side tokenization support for cleaned-random dataset

- Add SimplePromptClient and tokenization utilities to benchmark_utils.py
- Implement "cleaned-random" dataset option in benchmarks_serving_old.py
- Support server-side tokenization for more accurate token counting
- Maintain light-touch approach with minimal changes to original code
- Requires OPENAI_API_KEY environment variable and healthy vLLM server

This enables stable, cleaned random prompt generation using server-side
tokenization while preserving compatibility with existing benchmark code.

tstescoTT commented on benchmarks/benchmark_serving.py:
```diff
@@ -844,21 +897,21 @@ def main(args: argparse.Namespace):
-        "then all the requests are sent at time 0. "
-        "Otherwise, we use Poisson process or gamma distribution "
-        "to synthesize the request arrival times.",
+             "then all the requests are sent at time 0. "
```
Can you avoid whitespace diffs to the vLLM benchmark script? those will give diffs when we pull from upstream.

tstescoTT replied:
resolved

stisiTT committed:
auth_headers fix

stisiTT committed:
Whitespace fixes
